src/tools/r2rdump/R2RReader.cs

```diff
@@ -65,25 +75,164 @@ public unsafe R2RReader(string filename)
-                    }
-                }
-
+                // initialize the R2RHeader
                 DirectoryEntry r2rHeaderDirectory = peReader.PEHeaders.CorHeader.ManagedNativeHeaderDirectory;
-                int r2rHeaderOffset = r2rHeaderDirectory.RelativeVirtualAddress - textSection.VirtualAddress + textSection.PointerToRawData;
-                R2RHeader = new R2RHeader(_image, (uint)r2rHeaderDirectory.RelativeVirtualAddress, r2rHeaderOffset);
+                int r2rHeaderOffset = GetOffset(r2rHeaderDirectory.RelativeVirtualAddress);
+                R2RHeader = new R2RHeader(_image, r2rHeaderDirectory.RelativeVirtualAddress, r2rHeaderOffset);
                 if (r2rHeaderDirectory.Size != R2RHeader.Size)
                 {
                     throw new System.BadImageFormatException("The calculated size of the R2RHeader doesn't match the size saved in the ManagedNativeHeaderDirectory");
                 }
+
+                if (peReader.HasMetadata)
+                {
+                    var mdReader = peReader.GetMetadataReader();
+
+                    // initialize R2RMethods with method signatures from MethodDefHandle, and runtime function indices from MethodDefEntryPoints
+                    int i = 0;
+                    int methodDefEntryPointsRVA = R2RHeader.Sections[R2RSection.SectionType.READYTORUN_SECTION_METHODDEF_ENTRYPOINTS].RelativeVirtualAddress;
+                    int methodDefEntryPointsOffset = GetOffset(methodDefEntryPointsRVA);
+                    NativeArray methodEntryPoints = new NativeArray(_image, (uint)methodDefEntryPointsOffset);
+                    uint nMethods = methodEntryPoints.GetCount();
+                    R2RMethods = new R2RMethod[nMethods];
+                    for (uint rid = 1; rid <= nMethods; rid++)
+                    {
+                        uint offset = 0;
+                        if (methodEntryPoints.TryGetAt(_image, rid - 1, ref offset))
+                        {
+                            MethodDefinitionHandle methodDefHandle = MetadataTokens.MethodDefinitionHandle((int)rid);
+                            R2RMethods[i++] = new R2RMethod(_image, ref mdReader, methodDefHandle, methodEntryPoints, offset, rid);
+                        }
+                    }
+
+                    // get the RVAs of the runtime functions for each method
+                    int runtimeFunctionSize = 2;
+                    if (Machine == Machine.Amd64)
+                    {
+                        runtimeFunctionSize = 3;
+                    }
+                    R2RSection runtimeFunctionSection = R2RHeader.Sections[R2RSection.SectionType.READYTORUN_SECTION_RUNTIME_FUNCTIONS];
+                    int curOffset = GetOffset(runtimeFunctionSection.RelativeVirtualAddress);
+                    for (i = 0; i < nMethods; i++)
+                    {
+                        uint nRuntimeFunctions; // the number of runtime functions in this method
```

Comment:
I think this approach is good for now. I think later we can more easily extract the code length from the GCInfo, so keep in mind how this part of the code can be improved.